build/build_tarballs.jl

```diff
@@ -13,16 +17,84 @@ function create_tarball(dir, tarball)
     end
 end
 
-if abspath(PROGRAM_FILE) == @__FILE__
-    isdir(TARBALL_DIR) || mkdir(TARBALL_DIR)
+# https://docs.github.com/en/rest/actions/workflow-runs?apiVersion=2022-11-28#list-workflow-runs-for-a-repository
+function list_workflow_runs(; org, repo, head_sha)
+    # Needs to be a full SHA
+    url = "https://api.github.com/repos/$org/$repo/actions/runs?head_sha=$head_sha"
+    b = Downloads.download(url, IOBuffer())
+    return JSON3.read(seekstart(b))
+end
+
+# https://docs.github.com/en/rest/actions/artifacts?apiVersion=2022-11-28#list-workflow-run-artifacts
+function list_workflow_run_artifacts(; org, repo, run_id)
+    url = "https://api.github.com/repos/$org/$repo/actions/runs/$run_id/artifacts"
+    b = Downloads.download(url, IOBuffer())
+    return JSON3.read(seekstart(b))
+end
+
+function download_ray_julia_artifacts(; commit_sha, token, tarball_dir=TARBALL_DIR)
+    isdir(tarball_dir) || mkpath(tarball_dir)
+
+    response = list_workflow_runs(; org=REPO_ORG, repo=REPO_NAME, head_sha=commit_sha)
+    run_id = only(filter(r -> r.name == ARTIFACTS_WORKFLOW_NAME, response.workflow_runs)).id
+
+    response = list_workflow_run_artifacts(; org=REPO_ORG, repo=REPO_NAME, run_id)
+    artifacts = map(j -> j.name => j.archive_download_url, response.artifacts)
+
+    headers = ["Authorization" => "Bearer $token"]
+    for (name, url) in artifacts
+        startswith(name, "ray_julia") || continue
+
+        @info "Downloading $name"
+        io = Downloads.download(url, IOBuffer(); headers)
+        zip = ZipFile.Reader(io)
+
+        if length(zip.files) == 1
+            error("GitHub workflow artifact contains more than one file:\n$(join(zip.files, '\n'))")
+        end
+
+        file = only(zip.files)
+        write(joinpath(tarball_dir, name), read(file))
+    end
+
+    return nothing
+end
+
+function build_host_tarball(; tarball_dir=TARBALL_DIR, override::Bool=true)
+    isdir(tarball_dir) || mkdir(tarball_dir)
 
     @info "Building ray_julia library..."
-    include("build_library.jl")
+    build_library(; override)
 
     host = supported_platform(HostPlatform())
     tarball_name = gen_artifact_filename(; tag=TAG, platform=host)
 
     @info "Creating tarball $tarball_name"
-    tarball_path = joinpath(TARBALL_DIR, tarball_name)
+    tarball_path = joinpath(tarball_dir, tarball_name)
     create_tarball(COMPILED_DIR, tarball_path)
+
+    return tarball_path
+end
+
+function main()
+    if "--fetch" in ARGS
+        token = get(ENV, "GITHUB_TOKEN") do
+            Base.shred!(Base.getpass("GitHub PAT")) do s
+                return read(s, String)
+            end
+        end
+
+        @info "Retrieving CI built tarballs..."
+        commit_sha = git_head_sha()
+        download_ray_julia_artifacts(; commit_sha, token)
+    else
+        override = !("--no-override" in ARGS)
+        build_host_tarball(; override)
+    end
+
+    return nothing
+end
+
+if abspath(PROGRAM_FILE) == @__FILE__
+    main()
```

Comment:
do we even use `main`? I feel like this is two functions stacked in a trenchcoat?

Might be better to pull out `download_ray_julia_artifacts` + friends into a `download_artifacts.jl` script.
That way we have a clean separation of what it means to build vs download and we don't have to worry about the mode-switching behaviour of `--fetch` (which kinda smells) and `--override` having no effect if fetching (another smell).

Conceptually I think it's much easier if `build_tarballs.jl --fetch` == `download_tarballs.jl` and is run locally in the build steps, while the new `build_tarballs.jl` is what runs in the CI.

Reply:
I was worried about making too many scripts. However, if we did end up consolidating the scripts we'd probably combine a few more than what we have. Alright, I'll turn this into a new script.